fluent/src/bundle.js

```diff
@@ -170,74 +153,124 @@ export default class FluentBundle {
    *
-   *     // Returns 'Hello, name!' and `errors` is now:
-   *
-   *     [<ReferenceError: Unknown variable: name>]
-   *
-   * @param   {Object | string}    message
-   * @param   {Object | undefined} args
-   * @param   {Array}              errors
+   * @param   {string} id
+   * @param   {?Object} args
+   * @param   {?Array} errors
    * @returns {?string}
    */
-  format(message, args, errors) {
-    // optimize entities which are simple strings with no attributes
-    if (typeof message === "string") {
-      return this._transform(message);
+  format(id, args, errors = []) {
+    if (!this._messages.has(id)) {
+      errors.push(`Message not found: "${id}"`);
+      return undefined;
     }
 
-    // optimize entities with null values
-    if (message === null || message.value === null) {
-      return null;
-    }
+    let scope = this._createScope(args, errors);
+    let message = this._messages.get(id);
+    let value = message.resolveValue(scope);
+    return (value instanceof FluentNone) ? null : value;
+  }
 
-    // optimize simple-string entities with attributes
-    if (typeof message.value === "string") {
-      return this._transform(message.value);
-    }
+  /**
+   * Format the value of a message to a string. If the message has no value,
+   * return null.
+   *
+   *     let errors = [];
+   *     bundle.addMessages("hello = Hello, {$name}!");
+   *
+   *     let message = bundle.getMessage("hello");
+   *     bundle.formatValue(message, {name: "Jane"}, errors);
+   *     // → "Hello, Jane!" and `errors` is empty.
+   *
+   * @param   {FluentMessage} message
+   * @param   {?Object} args
+   * @param   {?Array} errors
+   * @returns {?string}
+   */
+  formatValue(message, args, errors) {
+    let scope = this._createScope(args, errors);
+    let value = message.resolveValue(scope);
+    return (value instanceof FluentNone) ? null : value;
+  }
+
+  /**
+   * Format an attribute of a message to a string. If the attribute does not
+   * exist, return undefined.
+   *
+   *     let errors = [];
+   *     bundle.addMessages(`
+   *     your-email =
+   *         .placeholder = Type your e-mail
+   *     `);
+   *
+   *     let message = bundle.getMessage("your-email");
+   *     bundle.formatAttribute(message, "placeholder", null, errors);
+   *     // → "Type your e-mail" and `errors` is empty.
+   *
+   * @param   {FluentMessage} message
+   * @param   {?Object} args
+   * @param   {?Array} errors
+   * @returns {?string}
+   */
+  formatAttribute(message, name, args, errors) {
+    let scope = this._createScope(args, errors);
+    let attribute = message.resolveAttribute(scope, name);
+    return (attribute instanceof FluentNone) ? undefined : attribute;
+  }
 
-    return resolve(this, args, message, errors);
+  _createScope(args, errors = []) {
+    return {
+      args, errors,
+      bundle: this,
+      dirty: this._dirty,
+      // TermReferences are resolved in a new scope.
+      insideTermReference: false,
```

Comment:
I struggled with that in Rust because of lifetimes. While the exact problem is very rust specific and doesn't warrant design changes, I kinda like what I ended up with - `locale_args` in the scope - https://github.com/projectfluent/fluent-rs/blob/master/fluent-bundle/src/resolve.rs#L244

Maybe worth considering here? Not a blocker of course.

Reply:
Nice! So `local_args` is both a store of arguments passed to the term _and_ a flag meaning _we're resolving a term now_. I like it!